src/Storage/Cookie.php

```diff
@@ -0,0 +1,205 @@
+        }
+
+        $this->cookiename = $cookiename;
+
+        $this->ttl = $ttl;
+    }
+
+    /**
+     * Parses any previous cookie and stores is internally
+     * @return void
+     */
+    protected function parseExistingCookie() {
+
+        if (is_array($this->cookievalues)) {
+            return;
+        }
+
+        if (empty($_COOKIE) || !isset($_COOKIE[$this->cookiename])) {
+            $this->cookievalues = [];
+            return;
+        }
+
+        $deserialized_cookie = json_decode($_COOKIE[$this->cookiename], true);
+
+        if (is_null($deserialized_cookie)) {
+            $this->cookievalues = [];
+            return;
+        }
+
+        $this->cookievalues = $deserialized_cookie;
+    }
+
+    /**
+     * Saves cookie with serialized test values
+     * @return bool
+     */
+    protected function saveCookie() {
+
+        $this->parseExistingCookie();
+        return setcookie($this->cookiename, json_encode($this->cookievalues), time() + $this->ttl, '/');
+    }
+
+    /**
+     * @inheritDoc
+     * @throws InvalidArgumentException
+     */
+    public function has($identifier) {
+
+        if (!is_string($identifier) || empty($identifier)) {
+            throw new InvalidArgumentException('Test identifier is invalid.');
+        }
+
+        $this->parseExistingCookie();
+
+        return isset($this->cookievalues[$identifier]);
+    }
+
+    /**
+     * @inheritDoc
+     * @throws InvalidArgumentException
+     */
+    public function get($identifier) {
+
+        if (!is_string($identifier) || empty($identifier)) {
+            throw new InvalidArgumentException('Test identifier is invalid.');
+        }
+
+        $this->parseExistingCookie();
+
+        if (!$this->has($identifier)) {
+            return null;
+        }
+
+        return $this->cookievalues[$identifier];
+    }
+
+    /**
+     * @inheritDoc
+     * @throws InvalidArgumentException
+     * @throws RuntimeException
+     */
+    public function set($identifier, $participation) {
+
+        if (!is_string($identifier) || empty($identifier)) {
+            throw new InvalidArgumentException('Test identifier is invalid.');
+        }
+
+        if (empty($participation)) {
+            throw new InvalidArgumentException('Participation name is invalid.');
+        }
+
+        if (headers_sent()) {
+            throw new RuntimeException('Headers have been sent. Cannot save cookie.');
+        }
+
+        $this->parseExistingCookie();
+
+        $this->cookievalues[$identifier] = $participation;
+
+        return $this->saveCookie();
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function all() {
+
+        $this->parseExistingCookie();
+        return $this->cookievalues;
+    }
+
+    /**
+     * @inheritDoc
+     * @throws InvalidArgumentException
+     * @throws RuntimeException
+     */
+    public function remove($identifier) {
+
+        if (!is_string($identifier) || empty($identifier)) {
```

Comment:
Can't you use the has method btw? Also in the cases above. See the Session storage